File: ios/BlePaymentKit/Sources/BlePaymentKit/BlePayloadDecoder.swift
import Foundation

public struct BlePayloadDecoder: Sendable {
    private let advertisementParser: AdvertisementPacketParser
    private let scanResponseParser: ScanResponseParser

    public init(advertisementParser: AdvertisementPacketParser = AdvertisementPacketParser(), scanResponseParser: ScanResponseParser = ScanResponseParser()) {
        self.advertisementParser = advertisementParser
        self.scanResponseParser = scanResponseParser
    }

    public func decode(input: BlePacketInput, config: BlePaymentConfig) -> Result<(AdvertisementPacket, ScanResponsePayload), BlePacketRejectReason> {
        guard let scanResponseData = input.scanResponseData else { return .failure(.missingRequiredField) }
        switch advertisementParser.parse(input.advertisementData, config: config) {
        case .failure(let reason): return .failure(reason)
        case .success(let advertisement):
            let manufacturerId = input.manufacturerId ?? config.manufacturerId
            switch scanResponseParser.parse(manufacturerId: manufacturerId, payload: scanResponseData, config: config) {
            case .failure(let reason): return .failure(reason)
            case .success(let scanResponse): return .success((advertisement, scanResponse))
            }
        }
    }
}

File: android/ble-payment-kit/src/main/kotlin/ru/paymentguide/blepaymentkit/BlePayloadDecoder.kt
package ru.paymentguide.blepaymentkit

class BlePayloadDecoder(
    private val advertisementPacketParser: AdvertisementPacketParser = AdvertisementPacketParser(),
    private val scanResponseParser: ScanResponseParser = ScanResponseParser(),
) {
    fun decode(input: BlePacketInput, config: BlePaymentConfig): Result<Pair<AdvertisementPacket, ScanResponsePayload>> = runCatching {
        val scanResponseData = input.scanResponseData ?: throw PacketRejected(BlePacketRejectReason.missingRequiredField)
        val advertisement = advertisementPacketParser.parse(input.advertisementData, config).getOrElse { throw it }
        val scanResponse = scanResponseParser.parse(input.manufacturerId ?: config.manufacturerId, scanResponseData, config).getOrElse { throw it }
        advertisement to scanResponse
    }
}
